File: medicine_app/interaction_timing.py
from __future__ import annotations

import re
import unicodedata
from datetime import date
from typing import Any, Mapping


_WITHIN_HOURS_RE = re.compile(r"(?P<hours>\d+)\s*시간\s*이내\s*병용금기", re.IGNORECASE)
_AFTER_COURSE_RE = re.compile(
    r"(?P<subject>[^,，.;。]{1,120}?)\s*투여\s*중\s*및\s*종료\s*후\s*"
    r"(?P<amount>\d+)\s*(?P<unit>시간|일|주)\s*간",
    re.IGNORECASE,
)
_POST_COURSE_MARKERS = ("종료 후", "중단 후", "중단한 직후", "중단 직후", "투여 중 및 종료 후")


def _tokens(value: Any) -> set[str]:
    text = unicodedata.normalize("NFKC", str(value or "")).casefold()
    return set(re.findall(r"[0-9a-z가-힣]+", text))


def _source_side(subject: str, left_ingredient: Any, right_ingredient: Any) -> str | None:
    subject_tokens = _tokens(subject)
    if not subject_tokens:
        return None
    left_tokens = _tokens(left_ingredient)
    right_tokens = _tokens(right_ingredient)
    left_match = bool(left_tokens) and (
        subject_tokens.issubset(left_tokens) or left_tokens.issubset(subject_tokens)
    )
    right_match = bool(right_tokens) and (
        subject_tokens.issubset(right_tokens) or right_tokens.issubset(subject_tokens)
    )
    if left_match == right_match:
        return None
    return "left" if left_match else "right"


def parse_interaction_timing(
    text: Any,
    left_ingredient: Any,
    right_ingredient: Any,
) -> dict[str, Any]:
    """Structure only timing language whose duration and direction are explicit.

    DUR interaction reasons contain both repeated machine-resolvable wording and
    free text. Unresolved post-stop wording is intentionally surfaced as
    ``not_evaluable`` rather than guessed or silently reduced to overlap-only.
    """
    source_text = " ".join(str(text or "").split())
    match = _WITHIN_HOURS_RE.search(source_text)
    if match:
        amount = int(match.group("hours"))
        return {
            "status": "structured",
            "kind": "minimum_separation",
            "hours": amount,
            "amount": amount,
            "unit": "시간",
            "direction": "symmetric",
            "source_text": source_text,
        }
    match = _AFTER_COURSE_RE.search(source_text)
    if match:
        amount = int(match.group("amount"))
        multiplier = {"시간": 1, "일": 24, "주": 7 * 24}[match.group("unit")]
        side = _source_side(match.group("subject"), left_ingredient, right_ingredient)
        if side:
            return {
                "status": "structured",
                "kind": "washout_after",
                "hours": amount * multiplier,
                "amount": amount,
                "unit": match.group("unit"),
                "source_side": side,
                "subject": match.group("subject").strip(),
                "source_text": source_text,
            }
        return {
            "status": "not_evaluable",
            "kind": "post_course_restriction",
            "reason": "washout source ingredient could not be resolved uniquely",
            "source_text": source_text,
        }
    if any(marker in source_text for marker in _POST_COURSE_MARKERS):
        return {
            "status": "not_evaluable",
            "kind": "post_course_restriction",
            "reason": "post-course restriction duration could not be resolved",
            "source_text": source_text,
        }
    return {"status": "default", "kind": "course_overlap", "source_text": source_text}


def _course(value: Mapping[str, Any]) -> tuple[date, date | None] | None:
    raw_start = value.get("start_date")
    if not raw_start:
        return None
    try:
        start = date.fromisoformat(str(raw_start))
        end = date.fromisoformat(str(value["end_date"])) if value.get("end_date") else None
        stopped = (
            date.fromisoformat(str(value["stopped_at"]))
            if not value.get("active", True) and value.get("stopped_at")
            else None
        )
    except (TypeError, ValueError):
        return None
    if end is not None and end < start:
        return None
    if stopped is not None:
        if stopped < start:
            return None
        end = stopped if end is None else min(end, stopped)
    return start, end


def courses_overlap(left: Mapping[str, Any], right: Mapping[str, Any]) -> bool | None:
    left_course = _course(left)
    right_course = _course(right)
    if left_course is None or right_course is None:
        return None
    left_start, left_end = left_course
    right_start, right_end = right_course
    if left_end is not None and left_end < right_start:
        return False
    if right_end is not None and right_end < left_start:
        return False
    return True


def _potential_gap_within_hours(
    first: tuple[date, date | None],
    second: tuple[date, date | None],
    hours: int,
) -> bool:
    first_start, first_end = first
    second_start, second_end = second
    if first_end is not None and first_end < second_start:
        calendar_days = (second_start - first_end).days
        return max(calendar_days - 1, 0) * 24 < hours
    if second_end is not None and second_end < first_start:
        calendar_days = (first_start - second_end).days
        return max(calendar_days - 1, 0) * 24 < hours
    return True


def interaction_timing_applies(
    timing: Mapping[str, Any],
    candidate_course: Mapping[str, Any],
    current_course: Mapping[str, Any],
    *,
    candidate_side: str,
) -> bool:
    """Return whether a matching DUR pair is temporally relevant.

    Prescription dates have day precision, not administration timestamps. For
    hour-based rules a boundary calendar day remains potentially inside the
    stated interval and therefore stays flagged.
    """
    overlap = courses_overlap(candidate_course, current_course)
    if overlap is None:
        return True
    if overlap:
        return True
    kind = timing.get("kind")
    if timing.get("status") == "not_evaluable" and kind == "post_course_restriction":
        return True
    if kind == "course_overlap":
        return False
    candidate = _course(candidate_course)
    current = _course(current_course)
    if candidate is None or current is None:
        return True
    if kind == "minimum_separation":
        return _potential_gap_within_hours(candidate, current, int(timing["hours"]))
    if kind == "washout_after":
        row_left = candidate if candidate_side == "left" else current
        row_right = current if candidate_side == "left" else candidate
        source = row_left if timing.get("source_side") == "left" else row_right
        target = row_right if timing.get("source_side") == "left" else row_left
        source_start, source_end = source
        target_start, target_end = target
        if source_end is None:
            return not (target_end is not None and target_end < source_start)
        if target_start > source_end:
            return _potential_gap_within_hours(source, target, int(timing["hours"]))
        return False
    return False


__all__ = ["courses_overlap", "interaction_timing_applies", "parse_interaction_timing"]

File: rust/medicine_core/src/interaction_timing.rs
use regex::Regex;
use rusqlite::Connection;
use serde_json::{json, Map, Value};
use std::collections::BTreeSet;
use std::sync::OnceLock;
use unicode_normalization::UnicodeNormalization;

use crate::reference_semantics::semantic_records;
use crate::safety_time::{course, courses_overlap, Course};

const SUPPORTED_RUNTIME_EVALUATORS: [&str; 2] = ["minimum_separation", "excluded_route"];
const POST_COURSE_MARKERS: [&str; 5] = [
    "종료 후",
    "중단 후",
    "중단한 직후",
    "중단 직후",
    "투여 중 및 종료 후",
];

pub(crate) fn remark_timing(
    con: &Connection,
    row: &Map<String, Value>,
    details: Option<&str>,
) -> Result<Value, ()> {
    let semantics = semantic_records(con, row)?;
    if let Some(semantic) = semantics
        .iter()
        .find(|semantic| text(semantic, "evaluator_kind") == Some("minimum_separation"))
    {
        let payload = semantic
            .get("structured_payload")
            .and_then(Value::as_object);
        let hours = payload
            .and_then(|payload| payload.get("hours"))
            .and_then(Value::as_i64)
            .unwrap_or(0);
        return Ok(json!({
            "status": "structured",
            "kind": "minimum_separation",
            "hours": hours,
            "amount": hours,
            "unit": "시간",
            "direction": payload
                .and_then(|payload| payload.get("direction"))
                .and_then(Value::as_str)
                .unwrap_or("symmetric"),
            "source_text": semantic.get("source_remark").cloned().unwrap_or(Value::Null),
        }));
    }

    if let Some(semantic) = semantics.iter().find(|semantic| {
        text(semantic, "semantic_role") == Some("applicability_condition")
            && text(semantic, "evaluation_mode") == Some("runtime_evaluable")
            && !SUPPORTED_RUNTIME_EVALUATORS
                .contains(&text(semantic, "evaluator_kind").unwrap_or(""))
            && text(semantic, "fallback_action") == Some("review_required")
    }) {
        return Ok(json!({
            "status": "not_evaluable",
            "kind": "unknown_contract_evaluator",
            "reason": "reference condition evaluator is not supported by this app version",
            "source_text": semantic.get("source_remark").cloned().unwrap_or(Value::Null),
        }));
    }

    if let Some(semantic) = semantics.iter().find(|semantic| {
        text(semantic, "semantic_role") == Some("applicability_condition")
            && text(semantic, "evaluation_mode") == Some("review_required")
            && text(semantic, "fallback_action") == Some("review_required")
    }) {
        return Ok(json!({
            "status": "not_evaluable",
            "kind": text(semantic, "runtime_error_kind").unwrap_or("review_required_condition"),
            "reason": "reference condition requires professional review",
            "source_text": semantic.get("source_remark").cloned().unwrap_or(Value::Null),
        }));
    }
    Ok(parse_timing(
        details.unwrap_or(""),
        canonical_ingredient(row),
        canonical_paired_ingredient(row),
    ))
}

pub(crate) fn interaction_timing_applies(
    timing: &Map<String, Value>,
    candidate_course: &Map<String, Value>,
    current_course: &Map<String, Value>,
    candidate_side: &str,
) -> bool {
    let candidate = course(candidate_course);
    let current = course(current_course);
    if candidate.is_some_and(|value| value.empty) || current.is_some_and(|value| value.empty) {
        return false;
    }
    let overlap = courses_overlap(candidate_course, current_course);
    if overlap.is_none() || overlap == Some(true) {
        return true;
    }
    if text(timing, "status") == Some("not_evaluable") {
        return true;
    }
    match text(timing, "kind") {
        Some("course_overlap") => false,
        Some("minimum_separation") => match (candidate, current) {
            (Some(candidate), Some(current)) => {
                potential_gap_within_hours(candidate, current, integer(timing, "hours"))
            }
            _ => true,
        },
        Some("washout_after") => match (candidate, current) {
            (Some(candidate), Some(current)) => washout_applies(
                timing,
                candidate,
                current,
                candidate_side,
                integer(timing, "hours"),
            ),
            _ => true,
        },
        _ => false,
    }
}

fn parse_timing(
    details: &str,
    left_ingredient: Option<&str>,
    right_ingredient: Option<&str>,
) -> Value {
    let source_text = details.split_whitespace().collect::<Vec<_>>().join(" ");
    if let Some(captures) = within_hours_regex().captures(&source_text) {
        let amount = captures
            .name("hours")
            .and_then(|value| value.as_str().parse::<i64>().ok())
            .unwrap_or(0);
        return json!({
            "status": "structured",
            "kind": "minimum_separation",
            "hours": amount,
            "amount": amount,
            "unit": "시간",
            "direction": "symmetric",
            "source_text": source_text,
        });
    }
    if let Some(captures) = after_course_regex().captures(&source_text) {
        let amount = captures
            .name("amount")
            .and_then(|value| value.as_str().parse::<i64>().ok())
            .unwrap_or(0);
        let unit = captures.name("unit").map_or("", |value| value.as_str());
        let hours = amount
            * match unit {
                "일" => 24,
                "주" => 7 * 24,
                _ => 1,
            };
        let subject = captures
            .name("subject")
            .map_or("", |value| value.as_str())
            .trim();
        if let Some(side) = source_side(subject, left_ingredient, right_ingredient) {
            return json!({
                "status": "structured",
                "kind": "washout_after",
                "hours": hours,
                "amount": amount,
                "unit": unit,
                "source_side": side,
                "subject": subject,
                "source_text": source_text,
            });
        }
        return json!({
            "status": "not_evaluable",
            "kind": "post_course_restriction",
            "reason": "washout source ingredient could not be resolved uniquely",
            "source_text": source_text,
        });
    }
    if POST_COURSE_MARKERS
        .iter()
        .any(|marker| source_text.contains(marker))
    {
        return json!({
            "status": "not_evaluable",
            "kind": "post_course_restriction",
            "reason": "post-course restriction duration could not be resolved",
            "source_text": source_text,
        });
    }
    json!({
        "status": "default",
        "kind": "course_overlap",
        "source_text": source_text,
    })
}

fn washout_applies(
    timing: &Map<String, Value>,
    candidate: Course,
    current: Course,
    candidate_side: &str,
    hours: i64,
) -> bool {
    let (left, right) = if candidate_side == "left" {
        (candidate, current)
    } else {
        (current, candidate)
    };
    let (source, target) = if text(timing, "source_side") == Some("left") {
        (left, right)
    } else {
        (right, left)
    };
    if source.end.is_none() {
        return target
            .end
            .is_none_or(|target_end| target_end >= source.start);
    }
    if target.start > source.end.expect("checked above") {
        return potential_gap_within_hours(source, target, hours);
    }
    false
}

fn potential_gap_within_hours(first: Course, second: Course, hours: i64) -> bool {
    if let Some(first_end) = first.end {
        if first_end < second.start {
            let calendar_days = (second.start - first_end).num_days();
            return (calendar_days - 1).max(0) * 24 < hours;
        }
    }
    if let Some(second_end) = second.end {
        if second_end < first.start {
            let calendar_days = (first.start - second_end).num_days();
            return (calendar_days - 1).max(0) * 24 < hours;
        }
    }
    true
}

fn source_side(
    subject: &str,
    left_ingredient: Option<&str>,
    right_ingredient: Option<&str>,
) -> Option<&'static str> {
    let subject_tokens = tokens(subject);
    if subject_tokens.is_empty() {
        return None;
    }
    let left_tokens = tokens(left_ingredient.unwrap_or(""));
    let right_tokens = tokens(right_ingredient.unwrap_or(""));
    let left_match = !left_tokens.is_empty()
        && (subject_tokens.is_subset(&left_tokens) || left_tokens.is_subset(&subject_tokens));
    let right_match = !right_tokens.is_empty()
        && (subject_tokens.is_subset(&right_tokens) || right_tokens.is_subset(&subject_tokens));
    if left_match == right_match {
        None
    } else if left_match {
        Some("left")
    } else {
        Some("right")
    }
}

fn tokens(value: &str) -> BTreeSet<String> {
    let normalized = value
        .nfkc()
        .flat_map(char::to_lowercase)
        .collect::<String>();
    token_regex()
        .find_iter(&normalized)
        .map(|token| token.as_str().to_owned())
        .collect()
}

fn canonical_ingredient(row: &Map<String, Value>) -> Option<&str> {
    text(row, "criterion_ingredient_name").or_else(|| text(row, "ingredient_name"))
}

fn canonical_paired_ingredient(row: &Map<String, Value>) -> Option<&str> {
    text(row, "criterion_paired_ingredient_name").or_else(|| text(row, "paired_ingredient_name"))
}

fn integer(row: &Map<String, Value>, key: &str) -> i64 {
    row.get(key).and_then(Value::as_i64).unwrap_or(0)
}

fn text<'a>(row: &'a Map<String, Value>, key: &str) -> Option<&'a str> {
    row.get(key).and_then(Value::as_str)
}

fn within_hours_regex() -> &'static Regex {
    static REGEX: OnceLock<Regex> = OnceLock::new();
    REGEX.get_or_init(|| {
        Regex::new(r"(?P<hours>\d+)\s*시간\s*이내\s*병용금기")
            .expect("valid interaction window regex")
    })
}

fn after_course_regex() -> &'static Regex {
    static REGEX: OnceLock<Regex> = OnceLock::new();
    REGEX.get_or_init(|| {
        Regex::new(
            r"(?P<subject>[^,，.;。]{1,120}?)\s*투여\s*중\s*및\s*종료\s*후\s*(?P<amount>\d+)\s*(?P<unit>시간|일|주)\s*간",
        )
        .expect("valid washout regex")
    })
}

fn token_regex() -> &'static Regex {
    static REGEX: OnceLock<Regex> = OnceLock::new();
    REGEX.get_or_init(|| Regex::new(r"[0-9a-z가-힣]+").expect("valid interaction token regex"))
}
